### PR TITLE
Bug fix in Block.cpp

### PR DESCRIPTION
**Pull request by ruck314**
### Description
- fixed the Block's retry behavior and error print statements

### CODE REVIEW

**Comment by ruck314** [comment]
Here's the old (`incorrect behavior`):
```
1657064772.379156:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 1 out of 6: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064772.379308:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 2 out of 6: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064772.379392:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 3 out of 6: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064772.379470:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 4 out of 6: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
ERROR:pyrogue.Variable.RemoteVariable.Root.Core.AxiMicronN25Q[0].PasswordLock:Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
```
Here's the new (`correct behavior`):
```
1657064996.493623:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 1 out of 5: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064996.493743:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 2 out of 5: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064996.493828:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 3 out of 5: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064996.493901:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 4 out of 5: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
1657064996.493974:pyrogue.memory.block.Root.Core.AxiMicronN25Q[0].PasswordLock: Error on try 5 out of 5: Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
ERROR:pyrogue.Variable.RemoteVariable.Root.Core.AxiMicronN25Q[0].PasswordLock:Block::checkTransaction: General Error: Transaction error for block Root.Core.AxiMicronN25Q[0].PasswordLock with address 0x000200fc. Error Non zero status message returned on fpga register bus in hardware: 0x3
```

**Comment by bengineerd** [comment]
I looked at this separately just last week, and concluded that the current implementation is confusing but correct.

If you set a **retry** count of 5, that means it will try 6 times, because the first time is a **try**, and only subsequent tries are **retries**.

So a retry count of 0 means it will try once. A retry count of 1 means it will try twice.

It's confusing, but comes down to silly english semantics.

**Comment by bengineerd** [comment]
Looking a bit more closely, I think the changes on lines 285 and 318 are correct. The caught exception will now be re-thrown on the final try. But the log message doesn't need to change.

File: src/rogue/interfaces/memory/Block.cpp [hunk 1]
```diff
@@ -282,8 +282,8 @@ void rim::Block::startTransaction(uint32_t type, bool forceWr, bool check, rim::
          count = retryCount_;
 
       } catch ( rogue::GeneralError err ) {
-         if ( (count+1) >= retryCount_ ) throw err;
-         bLog_->error("Error on try %" PRIu32 " out of %" PRIu32 ": %s", (count+1), (retryCount_+1), err.what());
+         if ( (count+1) > retryCount_ ) throw err;
+         bLog_->error("Error on try %" PRIu32 " out of %" PRIu32 ": %s", (count+1), (retryCount_), err.what());
          fWr = true; // Stale state is now lost
```
Comment (bengineerd) [inline src/rogue/interfaces/memory/Block.cpp:287]: Should this maybe be
```C++
bLog_->error("Error on try %" PRIu32 " out of %" PRIu32 ": %s", (count+1), (retryCount_), err.what());
fWr = true; // Stale state is now lost
if ( (count+1) > retryCount_ ) throw err;
```

I think we want to log the error and set the stale state regardless, and then only re-throw the exception if this is the final try. 

@slacrherbst what do you think?

**Review by bengineerd — APPROVED** [review]

**Pull request merged**